spring-cloud-gateway-server/src/main/java/org/springframework/cloud/gateway/filter/factory/cache/postprocessor/SetMaxAgeHeaderAfterCacheExchangeMutator.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2013-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gateway.filter.factory.cache.postprocessor;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.springframework.cloud.gateway.filter.factory.cache.CachedResponse;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.server.reactive.ServerHttpResponse;
+import org.springframework.web.server.ServerWebExchange;
+
+/**
+ * @author Marta Medio
+ * @author Ignacio Lozano
+ */
+public class SetMaxAgeHeaderAfterCacheExchangeMutator implements AfterCacheExchangeMutator {
+
+	private static final String MAX_AGE_PREFIX = "max-age=";
+
+	private final Duration configuredTimeToLive;
+
+	private final Clock clock;
+
+	public SetMaxAgeHeaderAfterCacheExchangeMutator(Duration configuredTimeToLive, Clock clock) {
+		this.configuredTimeToLive = configuredTimeToLive;
+		this.clock = clock;
+	}
+
+	@Override
+	public void accept(ServerWebExchange exchange, CachedResponse cachedResponse) {
+		ServerHttpResponse response = exchange.getResponse();
+		long calculatedMaxAgeInSeconds = calculateMaxAgeInSeconds(cachedResponse, configuredTimeToLive);
+		rewriteCacheControlMaxAge(response.getHeaders(), calculatedMaxAgeInSeconds);
+
+	}
+
+	private long calculateMaxAgeInSeconds(CachedResponse cachedResponse, Duration configuredTimeToLive) {
+		long maxAge;
+		if (configuredTimeToLive.getSeconds() == -1) {
+			maxAge = -1;
+		}
+		else {
+			maxAge = Math.max(0, configuredTimeToLive.minus(getElapsedTimeInSeconds(cachedResponse)).getSeconds());
+		}
+
+		return maxAge;
+	}
+
+	private Duration getElapsedTimeInSeconds(CachedResponse cachedResponse) {
+		return Duration.ofMillis(clock.millis() - cachedResponse.timestamp().getTime());
+	}
+
+	private static void rewriteCacheControlMaxAge(HttpHeaders headers, long seconds) {
+		boolean isMaxAgePresent = Optional.ofNullable(headers.get(HttpHeaders.CACHE_CONTROL))
```

Comment:
Remove use of `Optional` as flow contraol. Use either `HttpHeaders.containsKey` or `HttpHeaders.getCacheControl()`